### PR TITLE
don't infer scope for locals with scoped destructor, fix issue 17934

### PR DESCRIPTION
**Pull request by WalterBright**
Need to find out why this inference was done.

### CODE REVIEW

**Comment by WalterBright** [comment]
Looks like I did it https://github.com/dlang/dmd/pull/7284

**Comment by ZombineDev** [comment]
> don't infer scope for locals with scoped destructor

Can you clarify why you think we shouldn't do inference?

**Comment by WalterBright** [comment]
It turns out that `scope` on a destructor means it isn't leaking the data, which has nothing to do with implying the instance itself should be `scope`.

File: test/fail_compilation/retscope3.d [hunk 1]
```diff
@@ -60,7 +60,7 @@ fail_compilation/retscope3.d(3027): Error: scope variable `l` assigned to `elem`
 
 #line 3000
 
-struct List
+scope struct List
```
[the remark this answers is not in the document]
Comment (Geod24) [reply test/fail_compilation/retscope3.d]: Not to mention `scope` in the case of `class` makes sense. `scope struct`, I've absolutely no idea what it does.

[the remark this answers is not in the document]
Comment (WalterBright) [reply test/fail_compilation/retscope3.d]: > I've absolutely no idea what it does.

It means any local assigned a `scope struct` value is inferred to be `scope` as well.

[the remark this answers is not in the document]
Comment (WalterBright) [reply test/fail_compilation/retscope3.d]: @jinshil this is about scope for structs, not for classes.

[the remark this answers is not in the document]
Comment (JinShil) [reply test/fail_compilation/retscope3.d]: @WalterBright if this is a legitimate use of `scope` then it needs to be documented somewhere describing what it does, preferable with a few examples illustrating the semantics.  I think this is the first use of `scope` on a struct in our entire D ecosystem, and prior to this PR, I don't know if attributing a struct with `scope` ever did anything.

[the remark this answers is not in the document]
Comment (JinShil) [reply test/fail_compilation/retscope3.d]: Also, I don't like the inconsistency where using `scope class` has no meaning, but using `scope struct` does.  I would prefer better uniformity in the language.  I get the impression that we're designing by the seat of our pants here.  Why can't users just attribute `scope` at the usage site?  What's the compelling use case for making it a type constraint?

[the remark this answers is not in the document]
Comment (Geod24) [reply test/fail_compilation/retscope3.d]: Also, how does this fixes thing ? If one leaves `scope` out of the definition, the code compiles.

**Comment by Geod24** [comment]
Where are we on this ?

**Comment by WalterBright** [comment]
@Geod24 Waiting for someone to approve it.

**Comment by 12345swordy** [comment]
This PR needs rebasing.

**Comment by Geod24** [comment]
> recognize a `scope struct` as being a struct that locals declared as instances of the struct will be inferred as scope.

And that's where the failure of the design is. Instead of having transitive `scope`, we're going to have to make the *type definition* include `scope`. This is very obviously not composable.

**Comment by 12345swordy** [comment]
cc @RazvanN7

**Comment by RazvanN7** [comment]
This requires a changelog entry and a spec PR.

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
There are a few issues with this PR:

- Does this actually fix Issue 17934? If yes, this should be linked to the associated issue in bugzilla.
- It requires a changelog entry and a spec PR since there is no mention in the docs about scope structs.
- Why do we need to have `scope struct` definitions? Can't we just use scope as a type qualifier whenever we want that?

**Comment by RazvanN7** [comment]
ping @WalterBright

**Comment by dkorpel** [comment]
Superseded by https://github.com/dlang/dmd/pull/13926
Since `scope struct` is deprecated (as discussed in the March 2022 foundation meeting), fixing issue 17934 will require a different approach.

**Pull request closed without merging**